File: backend/internal/domain/blog.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// BlogStatus represents the status of a blog post
type BlogStatus string

const (
	BlogStatusDraft     BlogStatus = "draft"
	BlogStatusPublished BlogStatus = "published"
	BlogStatusArchived  BlogStatus = "archived"
)

// Blog represents a blog post
type Blog struct {
	ID              uuid.UUID    `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	Title           string       `gorm:"size:200;not null" json:"title"`
	Slug            string       `gorm:"size:250;uniqueIndex;not null" json:"slug"`
	Excerpt         *string      `gorm:"type:text" json:"excerpt,omitempty"`
	Content         string       `gorm:"type:text;not null" json:"content"`
	FeaturedImage   *string      `gorm:"size:500" json:"featured_image,omitempty"`
	MetaTitle       *string      `gorm:"size:200" json:"meta_title,omitempty"`
	MetaDescription *string      `gorm:"size:500" json:"meta_description,omitempty"`
	MetaKeywords    *string      `gorm:"size:500" json:"meta_keywords,omitempty"`
	AuthorID        uuid.UUID    `gorm:"type:uuid;not null;index" json:"author_id"`
	CategoryID      *uuid.UUID   `gorm:"type:uuid;index" json:"category_id,omitempty"`
	Status          BlogStatus   `gorm:"type:blog_status;not null;default:'draft'" json:"status"`
	PublishedAt     *time.Time   `json:"published_at,omitempty"`
	ViewCount       int          `gorm:"default:0" json:"view_count"`
	CreatedAt       time.Time    `json:"created_at"`
	UpdatedAt       time.Time    `json:"updated_at"`

	// Relationships
	Author   *User         `gorm:"foreignKey:AuthorID" json:"author,omitempty"`
	Category *BlogCategory `gorm:"foreignKey:CategoryID" json:"category,omitempty"`
	Tags     []BlogTag     `gorm:"many2many:blog_post_tags;" json:"tags,omitempty"`
}

// TableName specifies the table name for Blog
func (Blog) TableName() string {
	return "blogs"
}

// IsPublished returns true if the blog is published
func (b *Blog) IsPublished() bool {
	return b.Status == BlogStatusPublished
}

// BlogCategory represents a blog category
type BlogCategory struct {
	ID          uuid.UUID `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	Name        string    `gorm:"size:100;uniqueIndex;not null" json:"name"`
	Slug        string    `gorm:"size:100;uniqueIndex;not null" json:"slug"`
	Description *string   `gorm:"type:text" json:"description,omitempty"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// TableName specifies the table name for BlogCategory
func (BlogCategory) TableName() string {
	return "blog_categories"
}

// BlogTag represents a blog tag
type BlogTag struct {
	ID        uuid.UUID `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	Name      string    `gorm:"size:100;uniqueIndex;not null" json:"name"`
	Slug      string    `gorm:"size:100;uniqueIndex;not null" json:"slug"`
	CreatedAt time.Time `json:"created_at"`
}

// TableName specifies the table name for BlogTag
func (BlogTag) TableName() string {
	return "blog_tags"
}

// BlogPostTag represents the many-to-many junction table
type BlogPostTag struct {
	BlogID uuid.UUID `gorm:"type:uuid;primaryKey" json:"blog_id"`
	TagID  uuid.UUID `gorm:"type:uuid;primaryKey" json:"tag_id"`
}

// TableName specifies the table name for BlogPostTag
func (BlogPostTag) TableName() string {
	return "blog_post_tags"
}

// CreateBlogRequest represents the request to create a blog post
type CreateBlogRequest struct {
	Title           string      `json:"title" binding:"required,min=3,max=200"`
	Excerpt         *string     `json:"excerpt,omitempty"`
	Content         string      `json:"content" binding:"required,min=10"`
	FeaturedImage   *string     `json:"featured_image,omitempty"`
	MetaTitle       *string     `json:"meta_title,omitempty"`
	MetaDescription *string     `json:"meta_description,omitempty"`
	MetaKeywords    *string     `json:"meta_keywords,omitempty"`
	CategoryID      *uuid.UUID  `json:"category_id,omitempty"`
	Status          BlogStatus  `json:"status" binding:"required,oneof=draft published archived"`
	TagIDs          []uuid.UUID `json:"tag_ids,omitempty"`
}

// UpdateBlogRequest represents the request to update a blog post
type UpdateBlogRequest struct {
	Title           *string     `json:"title,omitempty" binding:"omitempty,min=3,max=200"`
	Excerpt         *string     `json:"excerpt,omitempty"`
	Content         *string     `json:"content,omitempty" binding:"omitempty,min=10"`
	FeaturedImage   *string     `json:"featured_image,omitempty"`
	MetaTitle       *string     `json:"meta_title,omitempty"`
	MetaDescription *string     `json:"meta_description,omitempty"`
	MetaKeywords    *string     `json:"meta_keywords,omitempty"`
	CategoryID      *uuid.UUID  `json:"category_id,omitempty"`
	Status          *BlogStatus `json:"status,omitempty" binding:"omitempty,oneof=draft published archived"`
	TagIDs          *[]uuid.UUID `json:"tag_ids,omitempty"`
}

// BlogListResponse represents a paginated list of blogs
type BlogListResponse struct {
	Blogs      []Blog `json:"blogs"`
	Total      int64  `json:"total"`
	Page       int    `json:"page"`
	PageSize   int    `json:"page_size"`
	TotalPages int    `json:"total_pages"`
}

// BlogFilters represents filters for listing blogs
type BlogFilters struct {
	Status     *BlogStatus `json:"status,omitempty"`
	CategoryID *uuid.UUID  `json:"category_id,omitempty"`
	AuthorID   *uuid.UUID  `json:"author_id,omitempty"`
	Search     *string     `json:"search,omitempty"`
	Page       int         `json:"page"`
	PageSize   int         `json:"page_size"`
	SortBy     string      `json:"sort_by"`
	SortOrder  string      `json:"sort_order"`
}

// CreateCategoryRequest represents the request to create a category
type CreateCategoryRequest struct {
	Name        string  `json:"name" binding:"required,min=2,max=100"`
	Description *string `json:"description,omitempty"`
}

// UpdateCategoryRequest represents the request to update a category
type UpdateCategoryRequest struct {
	Name        *string `json:"name,omitempty" binding:"omitempty,min=2,max=100"`
	Description *string `json:"description,omitempty"`
}

// CreateTagRequest represents the request to create a tag
type CreateTagRequest struct {
	Name string `json:"name" binding:"required,min=2,max=100"`
}

File: frontend/src/lib/api/blog.ts
import apiClient from './client'

// Types
export type BlogStatus = 'draft' | 'published' | 'archived'

export interface BlogAuthor {
  id: string
  first_name: string
  last_name: string
  email: string
}

export interface BlogCategory {
  id: string
  name: string
  slug: string
  description?: string
  created_at: string
  updated_at: string
}

export interface BlogTag {
  id: string
  name: string
  slug: string
  created_at: string
}

export interface Blog {
  id: string
  title: string
  slug: string
  excerpt?: string
  content: string
  featured_image?: string
  meta_title?: string
  meta_description?: string
  meta_keywords?: string
  author_id: string
  author?: BlogAuthor
  category_id?: string
  category?: BlogCategory
  tags?: BlogTag[]
  status: BlogStatus
  published_at?: string
  view_count: number
  created_at: string
  updated_at: string
}

export interface BlogListResponse {
  blogs: Blog[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface BlogFilters {
  page?: number
  page_size?: number
  status?: BlogStatus
  category_id?: string
  search?: string
  sort_by?: string
  sort_order?: 'ASC' | 'DESC'
}

export interface CreateBlogRequest {
  title: string
  excerpt?: string
  content: string
  featured_image?: string
  meta_title?: string
  meta_description?: string
  meta_keywords?: string
  category_id?: string
  tag_ids?: string[]
  status: BlogStatus
}

export interface UpdateBlogRequest {
  title?: string
  excerpt?: string
  content?: string
  featured_image?: string
  meta_title?: string
  meta_description?: string
  meta_keywords?: string
  category_id?: string
  tag_ids?: string[]
  status?: BlogStatus
}

export interface CreateCategoryRequest {
  name: string
  description?: string
}

export interface UpdateCategoryRequest {
  name?: string
  description?: string
}

export interface CreateTagRequest {
  name: string
}

// Blog Generation Types
export interface GenerateBlogRequest {
  url?: string
  prompt?: string // Optional if URL is provided
  category_id?: string
  target_tone?: string
  target_length?: string
}

export interface GeneratedBlogResponse {
  title: string
  slug: string
  excerpt: string
  content: string
  featured_image?: string
  meta_title: string
  meta_description: string
  meta_keywords: string
  suggested_tags?: string[]
  source_url?: string
  image_search_term?: string
}

export interface UnsplashImage {
  id: string
  url: string
  thumb_url: string
  download_url: string
  alt_text: string
  photographer: string
  profile_url: string
  width: number
  height: number
}

export interface BlogPreviewResponse {
  success: boolean
  generated_blog: GeneratedBlogResponse
  warnings?: string[]
  image_options?: UnsplashImage[]
}

export interface CreateFromGeneratedRequest {
  generated_data: GeneratedBlogResponse
  edits?: Record<string, unknown>
  category_id?: string
  tag_ids?: string[]
  status?: string
}

export interface SearchImagesRequest {
  query: string
  per_page?: number
}

export interface SearchImagesResponse {
  success: boolean
  images: UnsplashImage[]
  total: number
}

export interface SimplifyContentRequest {
  content: string
  target_tone?: string
}

export interface SimplifyContentResponse {
  success: boolean
  simplified_content: string
  original_length: number
  simplified_length: number
}

// Public API functions
export const blogApi = {
  getBlogs: async (filters?: BlogFilters): Promise<BlogListResponse> => {
    const params = new URLSearchParams()

    if (filters?.page) params.append('page', filters.page.toString())
    if (filters?.page_size) params.append('page_size', filters.page_size.toString())
    if (filters?.category_id) params.append('category_id', filters.category_id)
    if (filters?.search) params.append('search', filters.search)

    const response = await apiClient.get<BlogListResponse>(`/blogs?${params.toString()}`)
    return response.data
  },

  getBlog: async (id: string): Promise<Blog> => {
    const response = await apiClient.get<Blog>(`/blogs/${id}`)
    return response.data
  },

  getBlogBySlug: async (slug: string): Promise<Blog> => {
    const response = await apiClient.get<Blog>(`/blogs/slug/${slug}`)
    return response.data
  },

  getCategories: async (): Promise<BlogCategory[]> => {
    const response = await apiClient.get<BlogCategory[]>('/blog-categories')
    return response.data
  },

  getTags: async (): Promise<BlogTag[]> => {
    const response = await apiClient.get<BlogTag[]>('/blog-tags')
    return response.data
  },
}

// Admin API functions
export const adminBlogApi = {
  createBlog: async (data: CreateBlogRequest): Promise<Blog> => {
    const response = await apiClient.post<Blog>('/admin/blogs', data)
    return response.data
  },

  getAdminBlogs: async (filters?: BlogFilters): Promise<BlogListResponse> => {
    const params = new URLSearchParams()

    if (filters?.page) params.append('page', filters.page.toString())
    if (filters?.page_size) params.append('page_size', filters.page_size.toString())
    if (filters?.status) params.append('status', filters.status)
    if (filters?.category_id) params.append('category_id', filters.category_id)
    if (filters?.search) params.append('search', filters.search)
    if (filters?.sort_by) params.append('sort_by', filters.sort_by)
    if (filters?.sort_order) params.append('sort_order', filters.sort_order)

    const response = await apiClient.get<BlogListResponse>(
      `/admin/blogs?${params.toString()}`
    )
    return response.data
  },

  getAdminBlog: async (id: string): Promise<Blog> => {
    const response = await apiClient.get<Blog>(`/admin/blogs/${id}`)
    return response.data
  },

  updateBlog: async (id: string, data: UpdateBlogRequest): Promise<Blog> => {
    const response = await apiClient.put<Blog>(`/admin/blogs/${id}`, data)
    return response.data
  },

  deleteBlog: async (id: string): Promise<void> => {
    await apiClient.delete(`/admin/blogs/${id}`)
  },

  publishBlog: async (id: string): Promise<Blog> => {
    const response = await apiClient.post<Blog>(`/admin/blogs/${id}/publish`)
    return response.data
  },

  unpublishBlog: async (id: string): Promise<Blog> => {
    const response = await apiClient.post<Blog>(`/admin/blogs/${id}/unpublish`)
    return response.data
  },

  // Category management
  createCategory: async (data: CreateCategoryRequest): Promise<BlogCategory> => {
    const response = await apiClient.post<BlogCategory>('/admin/blog-categories', data)
    return response.data
  },

  updateCategory: async (id: string, data: UpdateCategoryRequest): Promise<BlogCategory> => {
    const response = await apiClient.put<BlogCategory>(`/admin/blog-categories/${id}`, data)
    return response.data
  },

  deleteCategory: async (id: string): Promise<void> => {
    await apiClient.delete(`/admin/blog-categories/${id}`)
  },

  // Tag management
  createTag: async (data: CreateTagRequest): Promise<BlogTag> => {
    const response = await apiClient.post<BlogTag>('/admin/blog-tags', data)
    return response.data
  },

  deleteTag: async (id: string): Promise<void> => {
    await apiClient.delete(`/admin/blog-tags/${id}`)
  },

  // AI Blog Generation
  generateBlogPreview: async (data: GenerateBlogRequest): Promise<BlogPreviewResponse> => {
    const response = await apiClient.post<BlogPreviewResponse>(
      '/admin/blogs/generate/preview',
      data,
      { timeout: 120000 } // 2 minute timeout for AI generation
    )
    return response.data
  },

  createFromGenerated: async (data: CreateFromGeneratedRequest): Promise<Blog> => {
    const response = await apiClient.post<Blog>('/admin/blogs/generate/create', data)
    return response.data
  },

  searchImages: async (data: SearchImagesRequest): Promise<SearchImagesResponse> => {
    const response = await apiClient.post<SearchImagesResponse>(
      '/admin/blogs/generate/images',
      data
    )
    return response.data
  },

  simplifyContent: async (data: SimplifyContentRequest): Promise<SimplifyContentResponse> => {
    const response = await apiClient.post<SimplifyContentResponse>(
      '/admin/blogs/generate/simplify',
      data,
      { timeout: 60000 } // 1 minute timeout
    )
    return response.data
  },
}
